### PR TITLE
FlipBar, Moments and other fixes

### LINKED ISSUE #156: Etabs_Toolkit: Bar releases fail to push to Etabs
Errors when pushing bar releases to Etabs. Needs investigating and fixing

### LINKED ISSUE #232: ETABS_Toolkit: ETABS AddByPoint Sometimes Flips Bars
#### Description:
If bars run right to left or top to bottom they appear to be automatically flipped upon creation. We need a way to manage this or modify it for following properties for things like releases.

#### How to replicate:
Create a bar that is right to left (ie starts at 10,0,0 and ends at 0,0,0) or top to bottom (0,10,0 to 0,0,0) with a release on one side, and the release is on the opposite side when created in ETABS. I and J joints are also flipped accordingly.

#### Expected behaviour:
Bar directionality maintained and releases etc on the correct side.

#### Test file(s):

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/ETABS_Toolkit/Issue232-EtabsFlippingBars?csf=1&e=pb8MNB

### LINKED ISSUE #239: ETABS_Toolkit: Moment not pushed!
#### Description:
Can't push moment to ETABS17.

#### How to replicate:
Try push a moment. use testfile.

#### Test file(s):
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/ETABS_Toolkit/Issue_239%20no%20moments/EtabsMomentTest.gh?csf=1&e=ygRxdj

### picture of code:
etabs no moment

### LINKED ISSUE #249: ETABS_Toolkit: ExplicitSection does not Push
#### Description:
`ExplicitSection` does not Push, which is a problem since that would be the fallback for users for unimplemented things.

#### How to replicate:

Push a `ExplicitSection`, and note that is does not arrive

#### Test file(s):

https://burohappold.sharepoint.com/:u:/s/BHoM/ET5SV2hQUEhLkMYFAe5M1aABrGVHDhrynb6DBKyaD_qZig?e=3BLfCH

### LINKED ISSUE #252: ETABS_Toolkit: Cutback/Offset not pushing
#### Description:

Offsets in the X-Axis does not push, which is not a huge issue since ETABS mainly works with autoEndOffsets but it is also a very small fix.

https://github.com/BHoM/ETABS_Toolkit/blob/99ec33fb29551f54dbe32d2448504b91f536cfd2/Etabs_Adapter/Create/Bar.cs#L135

this sets both offsets as negative (as the BHoM convetion is defined in the local coordinate system), they should both be positive

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR

Closes #156 
Fixed by adding warnings when pushing unstable releases. The warning outline why the release was unstable but does not stop the bar from being pushed. The release is then set as ETABS default fix fix.
Closes #232  
Fixed by flipping the bar before ETABS gets the chance to. (but actually handling the asymmetric properties)
Residual issue is that asymmetric sections can't be mirrored in ETABS API, raised in #253 
Closes #239 
Closes #249 
Closes #252 

 ### Test files

https://burohappold.sharepoint.com/:u:/s/BHoM/EQS3iMoVy-1NtG9g1powvwMBAHCMpXf9frWfX42K-OmwyQ?e=N9EnZI

 ### Changelog
- Fixes unpredictable bar behaviour due to ETABS flipping them upon creation
- Adds warnings when pushing unstable `Bar` `Releases`
- Adds feature to push `BarUniformlyDistributedLoad` with moments
- Fixes pushing `ExplicitSection`
- Fixes Pushing `Offsets` with displacement along the X-axis

 ### Additional comments

### CODE REVIEW

File: ETABS_Engine/Query/FlipBar.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+namespace BH.Engine.ETABS
+{
+    public static partial class Query
+    {
+
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static bool FlipBar(this Bar bar)
```
Comment (IsakNaslundBh) [inline ETABS_Engine/Query/FlipBar.cs]: Would rename this method to something like "CheckFlipBar" or similar. Would assume this method to modify the bar, not just check it.

File: Etabs_Adapter/CRUD/Create/Bar.cs [hunk 2]
```diff
@@ -199,7 +219,15 @@ private void SetSection(ConcreteSection section)
 
         private void SetSection(ExplicitSection section)
         {
-            m_model.PropFrame.SetGeneral(section.Name, section.Material.Name, section.CentreZ * 2, section.CentreY * 2, section.Area, section.Asy, section.Asz, section.J, section.Iy, section.Iz, section.Wply, section.Wplz, section.Wely, section.Wely, section.Rgy, section.Rgz);
+            m_model.PropFrame.SetGeneral(section.Name, section.Material.Name, section.Vz + section.Vpz, section.Vy + section.Vpy, section.Area, section.Asy, section.Asz, section.J, section.Iy, section.Iz, section.Wply, section.Wplz, section.Wely, section.Wely, section.Rgy, section.Rgz);
```
Comment (IsakNaslundBh) [inline Etabs_Adapter/CRUD/Create/Bar.cs]: Found some errors in here, not induced by you additions, but that need to be fixed.

Section modulus `S` should be relating to `Wel`, right now Wpl is used.
Plastic modulus `Z` should relate to `Wpl`, right now `Wel` is used. `Wely` is also used twice.

Need to double check directions here as well, to make sure we have the right conversions from Iy/Iz to I22/33 as Etabs deals with this a bit differently than most other packages. From example section:

image

I33 > I22 hence I33 = Iy in BHoM. and similarly S33 = Wely, Z33 = Wply and R33 = Rgy.

Same logic As2 = Asz and As3 = Asy

File: ETABS_Engine/Modify/FlipEndPoints.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static Bar FlipEndPoints(this Bar bar)
+        {
+            // Flip the endpoints
+            Bar clone = bar.Flip();
+
+            // Flip orientationAngle
+            if (!bar.IsVertical())
```
Comment (IsakNaslundBh) [inline ETABS_Engine/Modify/FlipEndPoints.cs]: Generally working really well in terms of getting maintained orientation out. Found one corner case for a vertical bar:

image

File: ETABS_Engine/Convert/ToCSI.cs [hunk 4]
```diff
@@ -85,6 +85,25 @@ public static void ToCSI(this BarRelease release, ref bool[] startRestraint, ref
             endSpring[3] = release.EndRelease.RotationalStiffnessX;
             endSpring[4] = release.EndRelease.RotationalStiffnessY;
             endSpring[5] = release.EndRelease.RotationalStiffnessZ;
+
+            bool[] startReleased = startRestraint.Zip(startSpring, (x, y) => x && y == 0).ToArray();
```
Comment (IsakNaslundBh) [inline ETABS_Engine/Convert/ToCSI.cs]: Should take this opportunity to double check the direction of the releases here. Especially Uy,Uz,Ry and Rz and make sure the convert from BHoM to Etabs 1-3-2 system is done correctly.

**Comment by kThorsager** [comment]
Completely right @IsakNaslundBh , releases have been fixed (should that be logged somewhere?) and Explicit section is actually fixed now, and the bars orientation behaves

**Pull request merged**